content/issue-1/minilsm.md

```diff
@@ -0,0 +1,125 @@
+space. But in an LSM Tree, all write operations, i.e., insertions, updates, deletions, are performed in somewhere else.
+These operations will be batched into SST (sorted string table) files and be written to the disk. Once written to the
+disk, the file will not be changed. These operations are applied lazily on disk with a special task called compaction.
+The compaction job will merge multiple SST files and remove unused data.
+
+This architectural design makes LSM tree easy to work with.
+
+1. Data are immutable on persistent storage, which means that it is easier to offload the background tasks (compaction)
+   to remote servers. It is also feasible to directly store and serve data from cloud-native storage systems like S3.
+2. An LSM tree can balance between read, write and space amplification by changing the compaction algorithm. The data
+   structure itself is super versatile and can be optimized for different workloads.
+
+In this tutorial, we will learn how to build an LSM-Tree-based storage engine in the Rust programming language.
+
+## Overview of LSM
+
+An LSM storage engine generally contains 3 parts:
+
+1. Write-ahead log to persist temporary data for recovery.
+2. SSTs on the disk for maintaining a tree structure.
+3. Mem-tables in memory for batching small writes.
+
+The storage engine generally provides the following interfaces:
+
+* `Put(key, value)`: store a key-value pair in the LSM tree.
+* `Delete(key)`: remove a key and its corresponding value.
+* `Get(key)`: get the value corresponding to a key.
+* `Scan(range)`: get a range of key-value pairs.
+
+To ensure persistence,
+
+* `Sync()`: ensure all the operations before `sync` are persisted to the disk.
+
+Some engines choose to combine `Put` and `Delete` into a single operation called `WriteBatch`, which accepts a batch
+of key value pairs.
+
+In this tutorial, we assume the LSM tree is using leveled compaction algorithm, which is commonly used in real-world
+systems.
+
+## Write Flow
+
+![](https://skyzh.github.io/mini-lsm/figures/lsm-tutorial/00-lsm-write-flow.svg)
+
+The write flow of LSM contains 4 steps:
+
+1. Write the key-value pair to write-ahead log, so that it can be recovered after the storage engine crashes.
+2. Write the key-value pair to memtable. After (1) and (2) completes, we can notify the user that the write operation
+   is completed.
+3. When a memtable is full, we will flush it to the disk as an SST file in the background.
+4. We will compact some files in some level into lower levels to maintain a good shape for the LSM tree, so that read
+   amplification is low.
+
+## Read Flow
+
+![](https://skyzh.github.io/mini-lsm/figures/lsm-tutorial/00-lsm-read-flow.svg)
+
+When we want to read a key,
+
+1. We will first probe all the memtables from latest to oldest.
+2. If the key is not found, we will then search the entire LSM tree containing SSTs to find the data.
+
+## Tutorial Overview
```

Comment:
This section needs to add some key Rust code examples to make the article look fuller.